File: php/test/AdminconversationEntityTest.php
<?php
declare(strict_types=1);

// Adminconversation entity test

require_once __DIR__ . '/../slack_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class AdminconversationEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = SlackSDK::test(null, null);
        $ent = $testsdk->Adminconversation(null);
        $this->assertNotNull($ent);
    }

    // Feature #4: the entity stream(action, ...) method runs the op pipeline
    // and yields result items. With the streaming feature active it yields the
    // feature's incremental output; otherwise it falls back to the materialised
    // list so stream always yields.
    public function test_stream(): void
    {
        $seed = [
            "entity" => [
                "adminconversation" => [
                    "s1" => ["id" => "s1"],
                    "s2" => ["id" => "s2"],
                    "s3" => ["id" => "s3"],
                ],
            ],
        ];

        // Fallback: streaming inactive -> yields the materialised list items.
        $base = SlackSDK::test($seed, null);
        $seen = iterator_to_array($base->Adminconversation(null)->stream("list", null, null), false);
        $this->assertCount(3, $seen);

        // Inbound: streaming active -> yields each item from the feature.
        $cfg = SlackConfig::shared_config();
        if (isset($cfg["feature"]) && is_array($cfg["feature"]) && isset($cfg["feature"]["streaming"])) {
            $sdk = SlackSDK::test($seed, ["feature" => ["streaming" => ["active" => true]]]);
            $got = [];
            foreach ($sdk->Adminconversation(null)->stream("list", null, null) as $item) {
                if (is_array($item) && array_is_list($item)) {
                    foreach ($item as $sub) {
                        $got[] = $sub;
                    }
                } else {
                    $got[] = $item;
                }
            }
            $this->assertCount(3, $got);
        }
    }

    public function test_basic_flow(): void
    {
        $setup = adminconversation_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create", "list", "load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "adminconversation." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set SLACK_TEST_ADMINCONVERSATION_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $adminconversation_ref01_ent = $client->Adminconversation(null);
        $adminconversation_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.adminconversation"), "adminconversation_ref01"));

        $adminconversation_ref01_data_result = $adminconversation_ref01_ent->create($adminconversation_ref01_data, null);
        $adminconversation_ref01_data = Helpers::to_map(is_object($adminconversation_ref01_data_result) && method_exists($adminconversation_ref01_data_result, 'data_get') ? $adminconversation_ref01_data_result->data_get() : $adminconversation_ref01_data_result);
        $this->assertNotNull($adminconversation_ref01_data);
        $this->assertNotNull($adminconversation_ref01_data["id"]);

        // LIST
        $adminconversation_ref01_match = [];

        $adminconversation_ref01_list_result = $adminconversation_ref01_ent->list($adminconversation_ref01_match, null);
        $this->assertIsArray($adminconversation_ref01_list_result);

        $found_item = sdk_select(
            Runner::entity_list_to_data($adminconversation_ref01_list_result),
            ["id" => $adminconversation_ref01_data["id"]]);
        $this->assertNotEmpty($found_item);

        // LOAD
        $adminconversation_ref01_match_dt0 = [
            "id" => $adminconversation_ref01_data["id"],
        ];
        $adminconversation_ref01_data_dt0_loaded = $adminconversation_ref01_ent->load($adminconversation_ref01_match_dt0, null);
        $adminconversation_ref01_data_dt0_load_result = Helpers::to_map(is_object($adminconversation_ref01_data_dt0_loaded) && method_exists($adminconversation_ref01_data_dt0_loaded, 'data_get') ? $adminconversation_ref01_data_dt0_loaded->data_get() : $adminconversation_ref01_data_dt0_loaded);
        $this->assertNotNull($adminconversation_ref01_data_dt0_load_result);
        $this->assertEquals($adminconversation_ref01_data_dt0_load_result["id"], $adminconversation_ref01_data["id"]);

    }
}

function adminconversation_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/adminconversation/AdminconversationTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = SlackSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["adminconversation01", "adminconversation02", "adminconversation03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("SLACK_TEST_ADMINCONVERSATION_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "SLACK_TEST_ADMINCONVERSATION_ENTID" => $idmap,
        "SLACK_TEST_LIVE" => "FALSE",
        "SLACK_TEST_EXPLAIN" => "FALSE",
        "SLACK_APIKEY" => "",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["SLACK_TEST_ADMINCONVERSATION_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["SLACK_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            // FIRST, so the generated fields below win: sdk-test-control.json's
            // test.client.options adds to the live client, it does not redirect it.
            Runner::live_client_options(),
            [
                "apikey" => $env["SLACK_APIKEY"],
            ],
            // ismap, not a plain "?? []" default: an empty PHP array is a
            // LIST, and a non-map later entry REPLACES the accumulated map in
            // merge - so the no-extras call discarded live_client_options()
            // and the apikey/server map above it.
            Vs::ismap($extra) ? $extra : new \stdClass(),
        ]);
        // "?? []" because merge legitimately answers with a stdClass when every
        // contributing entry is an EMPTY map - an SDK with no apikey and no
        // server variables generates an empty middle entry, so that is the
        // common case, not the edge one. to_map returns null for a non-array by
        // design, and the constructor takes a non-nullable array, so without the
        // fallback every such SDK died on "must be of type array, null given"
        // the moment live mode was switched on. Offline mode never reaches this
        // branch, which is why the offline suite stayed green.
        $client = new SlackSDK(Helpers::to_map($merged_opts) ?? []);
    }

    $live = $env["SLACK_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["SLACK_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}

File: go/test/adminconversation_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/slack-sdk/go"
	"github.com/voxgig-sdk/slack-sdk/go/core"

	vs "github.com/voxgig-sdk/slack-sdk/go/utility/struct"
)

func TestAdminconversationEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Adminconversation(nil)
		if ent == nil {
			t.Fatal("expected non-nil AdminconversationEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"adminconversation": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Adminconversation(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.SharedConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Adminconversation(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := adminconversationBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "adminconversation." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set SLACK_TEST_ADMINCONVERSATION_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		adminconversationRef01Ent := client.Adminconversation(nil)
		adminconversationRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "adminconversation"}), "adminconversation_ref01"))

		adminconversationRef01DataResult, err := adminconversationRef01Ent.Create(adminconversationRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		adminconversationRef01Data = core.ToMapAny(entityData(adminconversationRef01DataResult))
		if adminconversationRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if adminconversationRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		adminconversationRef01Match := map[string]any{}

		adminconversationRef01ListResult, err := adminconversationRef01Ent.List(adminconversationRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		adminconversationRef01List, adminconversationRef01ListOk := adminconversationRef01ListResult.([]any)
		if !adminconversationRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", adminconversationRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(adminconversationRef01List), map[string]any{"id": adminconversationRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// LOAD
		adminconversationRef01MatchDt0 := map[string]any{
			"id": adminconversationRef01Data["id"],
		}
		adminconversationRef01DataDt0Loaded, err := adminconversationRef01Ent.Load(adminconversationRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		adminconversationRef01DataDt0LoadResult := core.ToMapAny(entityData(adminconversationRef01DataDt0Loaded))
		if adminconversationRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if adminconversationRef01DataDt0LoadResult["id"] != adminconversationRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func adminconversationBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "adminconversation", "AdminconversationTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read adminconversation test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse adminconversation test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"adminconversation01", "adminconversation02", "adminconversation03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("SLACK_TEST_ADMINCONVERSATION_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"SLACK_TEST_ADMINCONVERSATION_ENTID": idmap,
		"SLACK_TEST_LIVE":      "FALSE",
		"SLACK_TEST_EXPLAIN":   "FALSE",
		"SLACK_APIKEY":         "",
	})

	idmapResolved := core.ToMapAny(env["SLACK_TEST_ADMINCONVERSATION_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["SLACK_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["SLACK_APIKEY"],
			},
			extraOpts,
		})
		client = sdk.NewSlackSDK(core.ToMapAny(mergedOpts))
	}

	live := env["SLACK_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["SLACK_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
